File: 2-Add-Two-Number/solution1.java

  Definition for singly-linked list.
  public class ListNode {
      int val;
      ListNode next;
      ListNode() {}
      ListNode(int val) { this.val = val; }
      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
  }
 
class Solution {
    
   
       Overview 
   
   
          Time 2ms (99.44%)
           Memory 47.8MB(48.72%)
   
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(0);
        ListNode temp = res;
        
        boolean offset = false;
        while(l1 != null  l2 != null){
            int sum = 0;
            if(l1 != null){
                sum += l1.val;
                l1 = l1.next;
            }
            if(l2 != null){
                sum += l2.val;
                l2 = l2.next;
            }
            if(offset){
                sum += 1;
                offset = false;
            }
            if(sum  9){
                offset = true;
                sum -= 10;
            }
            temp.next = new ListNode(sum);
            temp = temp.next;
        }
        
        if(offset){
            temp.next = new ListNode(1);
        }
        return res.next;
        
    }
    
    
}

File: 21-Merge-Two-Sorted-Lists/solution1.py
'''
Use the normal recursion way

However it is slow

Time: 75ms 36.54%
Memory: 13.9MB 32.71%
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        def merge(l1,l2,res):
            if l1 is None:
                if l2 is None:
                    return
                else:
                    res.next = l2
                    return
            if l2 is None:
                if l1 is None:
                    return
                else:
                    res.next = l1
                    return
            if l1.val < l2.val:
                res.next = l1
                merge(l1.next,l2,res.next)
            else:
                res.next = l2
                merge(l1,l2.next,res.next)
        
        res = ListNode(0,None)
        merge(list1,list2,res)
        return res.next
